### PR TITLE
feature: better frameworks like cli

### PR DESCRIPTION
**Pull request by vasucp1207**
Method like other frameworks use(mainly **vue**).

### Cli including,
- Dynamically project name.
- Choose from different default templates.
- Overriding the directory if it exits after confirmation.
- Validates package name.

Should try to commit in https://github.com/dai-shi/waku/pull/52 but I messed up and also thought it's a completely new strategy so I raised this PR instead.

### CODE REVIEW

**Comment by vasucp1207** [comment]
Regarding https://github.com/dai-shi/waku/pull/89 , templates only contain necessary files it does not include folders like node_modules etc. Well if you run ``npm run build`` a resulted ``out.cjs`` file is generated, but I can't figure out now how to publish it to npm to actually check it. Do you have any idea about this.

**Comment by dai-shi** [comment]
> Regarding #89 , templates only contain necessary files it does not include folders like node_modules etc.

As I understand, #89's request is just to put `.gitignore` file in those templates.
But, on second thought, I'm not sure what should be in there.
Does create-vue generate `.gitignore` file?

File: contrib/create-waku/cli.js [hunk 1]
```diff
@@ -1,46 +1,148 @@
 #!/usr/bin/env node
 
-const path = require("node:path");
-const fs = require("node:fs");
-const https = require("node:https");
+import fs from 'fs'
+import path from 'path'
```
Comment (dai-shi) [inline contrib/create-waku/cli.js:4]:
```suggestion
import fs from 'node:fs'
import path from 'node:path'
```

File: contrib/create-waku/cli.js [hunk 2]
```diff
@@ -1,46 +1,148 @@
...
   }
-});
+  console.log(`${bold(green(commandsMap.install[packageManager]))}`)
+  console.log(`${bold(green(commandsMap.dev[packageManager]))}`)
+  console.log()
+}
+
+init().catch((e) => {
+  console.error(e)
+})
```
Comment (dai-shi) [inline contrib/create-waku/cli.js:148]: Let's refactor those files in TypeScript after #64. cc @himself65

File: contrib/create-waku/package.json [hunk 3]
```diff
@@ -1,5 +1,15 @@
 {
   "name": "create-waku",
   "version": "0.4.10",
-  "bin": "./cli.js"
+  "bin": "./cli.js",
+  "type": "module",
+  "scripts": {
+    "build": "./node_modules/.bin/esbuild cli.js --bundle --platform=node --outfile=out.cjs",
```
Comment (dai-shi) [inline contrib/create-waku/package.json:7]: wait, you want to build it before publish? we don't have CD yet.

File: contrib/create-waku/package.json [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
+    "test": "node ./tests/generateTemplates.test.js"
+  },
+  "dependencies": {
+    "esbuild": "0.18.17",
```
Comment (dai-shi) [inline contrib/create-waku/package.json:11]: it should be in devDependencies.

**Review by dai-shi — CHANGES_REQUESTED** [review]
One big issue for me is now we have to maintain both examples and templates.
As we are adding the "build" process, we should copy files from examples folder, instead of copying them.

File: contrib/create-waku/cli.js (hunk 2 above)
Comment (dai-shi) [reply contrib/create-waku/cli.js:148]: It seems this adds the build process anyway, so let's make code in TypeScript.

File: contrib/create-waku/package.json (hunk 4 above)
Comment (dai-shi) [reply contrib/create-waku/package.json:11]: Can we use swc instead of esbuild?

File: contrib/playground/README.md [hunk 5]
```diff
@@ -0,0 +1 @@
+## For testing purpose
```
Comment (dai-shi) [inline contrib/playground/README.md:1]: please remove this.

File: contrib/create-waku/package.json (hunk 3 above)
Comment (vasucp1207) [reply contrib/create-waku/package.json]: Remove esbuild

**Comment by vasucp1207** [comment]
> > Regarding #89 , templates only contain necessary files it does not include folders like node_modules etc.
> 
> As I understand, #89's request is just to put `.gitignore` file in those templates. But, on second thought, I'm not sure what should be in there. Does create-vue generate `.gitignore` file?

Yes, create-vue generates a `.gitignore` file, I think a minimal file is okay for this like [create-vue .gitignore](https://github.com/vuejs/create-vue/blob/9bba10e2ff2dba6acfc130752b74addc0f99b1e6/template/base/_gitignore#L4)

File: contrib/create-waku/cli.js [hunk 6]
```diff
@@ -1,46 +1,148 @@
...
+  return !fs.existsSync(dir) || fs.readdirSync(dir).length === 0
+}
+
+async function init() {
+  const cwd = process.cwd();
+
+  let targetDir;
+  let defaultProjectName = 'waku-project'
+
+  const CHOICES = fs.readdirSync(path.resolve(cwd, '../../examples'))
```
[the remark this answers is not in the document]
Comment (vasucp1207) [reply contrib/create-waku/cli.js]: > This doesn't work as we don't publish examples dir.
> 
> We have two options.
> 
> 1. add build step to copy examples dir
> 2. download from github
> 
> Eventually, we should go with 1 because it fixes #81, but for now 2 is also fine. If we were to go with 1, we would like to make the code in TypeScript and transpile with SWC in the build step.

Summary,
- If we ``npm run build:esbuild`` an ``out.js`` file generated.
- Running ``npm run build:estemplate`` generates an ``copy.js`` file.
- If we run ``node copy.js`` templates copied from ``example`` dir and a ``template`` folder generated.
- Finally, running ``node cli.js`` works fine.

Problem with transpiling with ``swr``,
- If we transpiling files with ``swr`` then the ``imports`` in output files are not working ex,
```js
import { renderTemplate } from './renderTemplate'
```
This import statement would not be able to resolve the file.

Is, I'm missing something about this ``swr``.

[the remark this answers is not in the document]
Comment (dai-shi) [reply contrib/create-waku/cli.js]: Okay, I don't mind using esbuild for now. I just thought SWC might work better and we also use it in `waku`. We can try that later.

Can you move TS files into `./src` folder please?

[the remark this answers is not in the document]
Comment (vasucp1207) [reply contrib/create-waku/cli.js]: Yeah, sure we can look around it in the near future.

File: contrib/create-waku/package.json [hunk 7]
```diff
@@ -1,5 +1,22 @@
 {
   "name": "create-waku",
   "version": "0.4.10",
-  "bin": "./cli.js"
+  "bin": "out.js",
+  "type": "commonjs",
+  "scripts": {
```
Comment (dai-shi) [inline contrib/create-waku/package.json:6]: Let's add this:

```suggestion
  "files": [
    "src",
    "dist"
  ],
  "scripts": {
```

File: contrib/create-waku/package.json [hunk 8]
```diff
@@ -1,5 +1,22 @@
 {
   "name": "create-waku",
   "version": "0.4.10",
-  "bin": "./cli.js"
+  "bin": "out.js",
+  "type": "commonjs",
```
Comment (dai-shi) [inline contrib/create-waku/package.json:5]: I thought your first proposition was "module". Which do you prefer?
Ah, I see it's because esbuild outputs CJS.

File: contrib/create-waku/package.json [hunk 9]
```diff
@@ -1,5 +1,22 @@
 {
   "name": "create-waku",
   "version": "0.4.10",
-  "bin": "./cli.js"
+  "bin": "out.js",
```
Comment (dai-shi) [inline contrib/create-waku/package.json:4]: can you make it `dist/cli.js`?

File: contrib/create-waku/tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+    "target": "ESNext",
+    "moduleResolution": "node",
+    "allowSyntheticDefaultImports": true,
+    "outDir": "dist",
+    "rootDir": "./",
+    "strict": true,
+  },
+  "include": ["./"],
+  "exclude": ["node_modules", "dist"]
+}
```
Comment (dai-shi) [inline contrib/create-waku/tsconfig.json:12]: Let's make it consistent with the project config. we use module=nodenext for example.
Or, can we simply remove this file?

File: contrib/create-waku/src/renderTemplate.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+      renderTemplate(path.resolve(src, file), path.resolve(dest, file));
+    }
+    return;
+  }
+
+  const filename = path.basename(src);
+
+  if (filename === 'package.json' && fs.existsSync(dest)) {
+    // merge instead of overwriting
+    const pkg = deepMerge(
```
Comment (dai-shi) [inline contrib/create-waku/src/renderTemplate.ts:20]: why do we need deep merge here?

File: contrib/create-waku/src/renderTemplate.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+      // @ts-expect-error
+      JSON.parse(fs.readFileSync(dest)), JSON.parse(fs.readFileSync(src))
+    );
+    fs.writeFileSync(dest, JSON.stringify(pkg, null, 2) + '\n');
+    return;
+  }
+
+  if (filename.startsWith('_')) {
+    // rename `_file` to `.file`
```
Comment (dai-shi) [inline contrib/create-waku/src/renderTemplate.ts:29]: Can we simply have dot files?

**Review by dai-shi — COMMENTED** [review]
Thanks for working on it. Here are some reviews.

File: contrib/create-waku/package.json (hunk 7 above)
Comment (vasucp1207) [reply contrib/create-waku/package.json:6]: Should we also include ``template`` in this?

File: contrib/create-waku/tsconfig.json (hunk 10 above)
Comment (vasucp1207) [reply contrib/create-waku/tsconfig.json:12]: Yeah, removing this is fine for now.

File: contrib/create-waku/src/renderTemplate.ts (hunk 11 above)
Comment (vasucp1207) [reply contrib/create-waku/src/renderTemplate.ts:20]: For using ``projectName`` as ``name`` key in ``package.json`` we created the ``package.json`` already, so here we have to deep merge rest of the fields.

File: contrib/create-waku/package.json (hunk 7 above)
Comment (dai-shi) [reply contrib/create-waku/package.json:10]: Oh, good point. We need that too.

File: contrib/create-waku/src/renderTemplate.ts (hunk 11 above)
Comment (dai-shi) [reply contrib/create-waku/src/renderTemplate.ts:20]: Can't it be shallow merge??

File: contrib/create-waku/src/cli.ts [hunk 13]
```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env node
+
+import fs from 'node:fs'
+import path from 'node:path'
+import prompts from 'prompts'
+import { red, green, bold } from 'kolorist'
+
+import { emptyDir } from './emptyDir'
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:8]: We need an extension for module=nodenext. `.js` not `.ts`.
```suggestion
import { emptyDir } from './emptyDir.js'
```

File: contrib/create-waku/src/renderTemplate.ts (hunk 11 above)
Comment (vasucp1207) [reply contrib/create-waku/src/renderTemplate.ts]: Yes, its working.

File: contrib/create-waku/src/cli.ts (hunk 13 above)
Comment (vasucp1207) [reply contrib/create-waku/src/cli.ts:8]: Do you mean adding ``.js`` to all importing files?

Comment (dai-shi) [reply contrib/create-waku/src/cli.ts:8]: https://www.typescriptlang.org/docs/handbook/esm-node.html

> As a result, it will have to be rewritten to use the extension of the output of foo.ts - so bar.ts will instead have to import from ./foo.js.

File: contrib/create-waku/src/cli.ts [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+        ],
+      }
+    ], {
+      onCancel: () => {
+        throw new Error(red('✖') + ' Operation cancelled')
+      }
+    })
+  } catch (cancelled) {
+    if(cancelled instanceof Error) 
+      console.log(cancelled.message)
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:91]: Just a preference:
```suggestion
    if(cancelled instanceof Error) {
      console.log(cancelled.message)
    }
```

File: contrib/create-waku/src/cli.ts [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+
+  const pkg = { name: packageName?? toValidPackageName(targetDir), version: '0.0.0' }
+
+  fs.writeFileSync(
+    path.resolve(root, 'package.json'),
+    JSON.stringify(pkg, null, 2),
+  )
+
+  console.log("Setting up project...")
+  const templateRoot = path.resolve(cwd, 'template')
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:113]: I think `path.resolve('template')` is the same. But, not super important anyway.

File: contrib/create-waku/src/cli.ts [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+    console.log(`${bold(green(`cd ${path.relative(cwd, root)}`))}`)
+  }
+  console.log(`${bold(green(commandsMap.install[packageManager]))}`)
+  console.log(`${bold(green(commandsMap.dev[packageManager]))}`)
+  console.log()
+}
+
+init().catch((e) => {
+  console.error(e)
+})
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:153]: This file feels a little bit complicated for me. Maybe, we look for something in the future.

File: contrib/create-waku/src/copyTemplateDir.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+  let root = path.join(cwd, '/template')
+  
+  if(!fs.existsSync(root)) {
+    fs.mkdirSync(root)
+  }
+  
+  const CHOICES = fs.readdirSync(path.resolve(cwd, '../../examples'))
+  const templateRoot = path.resolve(cwd, '../../examples')
+  
+  for(let i = 0; i < 3; i++) {
```
Comment (dai-shi) [inline contrib/create-waku/src/copyTemplateDir.ts:16]: Can `3` be `CHOICES.length`?

File: contrib/create-waku/src/renderTemplate.ts [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+      renderTemplate(path.resolve(src, file), path.resolve(dest, file));
+    }
+    return;
+  }
+
+  const filename = path.basename(src);
+
+  if (filename === 'package.json' && fs.existsSync(dest)) {
+    // merge instead of overwriting
+    const pkg = shallowMerge(
```
Comment (dai-shi) [inline contrib/create-waku/src/renderTemplate.ts:20]: Can shallow merging simply be `Object.assign({}, obj1, obj2)`?

File: contrib/create-waku/src/shallowMerge.ts [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+
+export function shallowMerge(target: any, obj: any) {
+  for (const key of Object.keys(obj)) {
+    const oldVal = target[key]
+    const newVal = obj[key]
+
+    if(key === 'name' || key === 'version') continue
+
+    if (Array.isArray(oldVal) && Array.isArray(newVal)) {
+      target[key] = mergeArrayWithDedupe(oldVal, newVal)
```
Comment (dai-shi) [inline contrib/create-waku/src/shallowMerge.ts:11]: ah, you want to deep merge arrays.

**Review by dai-shi — COMMENTED** [review]
tbh, there are many things I'd like to change, like coding practice, naming, structuring.
I'd like to move this forward, but in order for me to refactor things, tests might be important.
If you have any ideas to simplify things for now, please let me know.

File: contrib/create-waku/src/renderTemplate.ts [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+    fs.mkdirSync(dest, { recursive: true });
+    for (const file of fs.readdirSync(src)) {
+      renderTemplate(path.resolve(src, file), path.resolve(dest, file));
+    }
+    return;
+  }
+
+  const filename = path.basename(src);
+
+  if (filename === 'package.json' && fs.existsSync(dest)) {
```
Comment (dai-shi) [inline contrib/create-waku/src/renderTemplate.ts:17]: Can you explain again when this happens?

**Review by dai-shi — COMMENTED** [review]
Getting good. I still have many requests. Please be patient.

File: contrib/create-waku/src/cli.ts [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+
+  if (shouldOverwrite) {
+    emptyDir(root)
+  } else if (!fs.existsSync(root)) {
+    fs.mkdirSync(root)
+  }
+
+  const pkg = { name: packageName?? toValidPackageName(targetDir), version: '0.0.0' }
+
+  fs.writeFileSync(
```
Comment (vasucp1207) [inline contrib/create-waku/src/cli.ts:108]: Here, ``package.json`` already created with the input project name.

File: contrib/create-waku/src/renderTemplate.ts (hunk 20 above)
Comment (vasucp1207) [reply contrib/create-waku/src/renderTemplate.ts:17]: Maybe ``fs.existsSync(dest)``is not much relevant here.

File: contrib/create-waku/src/cli.ts (hunk 21 above)
Comment (dai-shi) [reply contrib/create-waku/src/cli.ts:108]: I mean `copyTemplateDir.ts` implementation can be much simplified because it doesn't need to "render".

Comment (dai-shi) [reply contrib/create-waku/src/cli.ts:108]: So, I don't think we need `renderTempalte.ts` at all (it doesn't mean to just move code).
copyTemplateDir is basically `cp -r src dest` and we could implement it easily, or can use some npm packages.
In cli.ts, we copy from a template and "edit" `package.json`. We don't need to do it during copy. I guess it simplifies the logic. As we start using some dependencies, we could use something for it too.

Basically, we want to reduce our code.

Comment (dai-shi) [reply contrib/create-waku/src/cli.ts:106]: Thanks, that's cool. Feels more maintainable.

File: contrib/create-waku/package.json [hunk 22]
```diff
@@ -1,5 +1,27 @@
...
+  "files": [
+    "src",
+    "dist",
+    "template"
+  ],
+  "type": "commonjs",
+  "scripts": {
+    "start": "node dist/cli.js",
+    "template": "cp -r ../../examples template/",
+    "test": "node dist/generateTemplates.test.js",
```
Comment (dai-shi) [inline contrib/create-waku/package.json:14]: How and when do I use this?

File: contrib/create-waku/src/cli.ts [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+    const newPackageJson = JSON.parse(fs.readFileSync(newPackageJsonPath, 'utf-8'))
+
+    const mergedPackageJson = {
+      ...newPackageJson,
+      ...existingPackageJson
+    }
+
+    const result = spawnSync('cp', ['-r', `${templateDir}/`, root])
+
+    // Write the merged package.json back to the target directory
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:132]: Nice.

File: contrib/create-waku/src/cli.ts [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+  return !fs.existsSync(dir) || fs.readdirSync(dir).length === 0
+}
+
+async function init() {
+  const cwd = process.cwd()
+
+  let targetDir = '';
+  let defaultProjectName = 'waku-project'
+
+  const CHOICES = fs.readdirSync(path.resolve(cwd, 'template'))
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:35]: I'm curious if we can simplify this, because cwd is always `process.cwd()`.

```suggestion
  const CHOICES = fs.readdirSync('template')
```

File: contrib/create-waku/src/cli.ts [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+    // Read new package.json from the template directory
+    const newPackageJsonPath = path.join(templateDir, 'package.json')
+    const newPackageJson = JSON.parse(fs.readFileSync(newPackageJsonPath, 'utf-8'))
+
+    const mergedPackageJson = {
+      ...newPackageJson,
+      ...existingPackageJson
+    }
+
+    const result = spawnSync('cp', ['-r', `${templateDir}/`, root])
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:130]: This probably doesn't work on Windows.
Let's use some libraries. I often  use https://www.npmjs.com/package/shelljs, but there can be some others that fit well with this use case.

**Review by dai-shi — COMMENTED** [review]
I like the new approach.

**Comment by dai-shi** [comment]
Please also take care of CI errors.

File: contrib/create-waku/package.json (hunk 22 above)
Comment (vasucp1207) [reply contrib/create-waku/package.json:14]: Copy templates with ``npm run template`` before publishing the package, ``npm run start`` is just for running the cli locally.

Comment (dai-shi) [reply contrib/create-waku/package.json:14]: should we run `npm test` after `npm run template`?

File: contrib/create-waku/src/cli.ts [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    process.exit(1);
+  }
+
+  const { packageName, shouldOverwrite, chooseProject } = result;
+
+  const root = path.join(cwd, targetDir);
+
+  if (shouldOverwrite) {
+    spawn.sync("rm", ["-rf", `${root}/*`]);
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:104]: Interesting. So, cross-spawn translates it for windows? (or does windows support `rm -rf`?)

**Review by dai-shi — CHANGES_REQUESTED** [review]
I'll take another look later. Please check CI errors meanwhile.

File: contrib/create-waku/src/cli.ts (hunk 26 above)
Comment (vasucp1207) [reply contrib/create-waku/src/cli.ts]: Updated with ``fs-extra`` node also have ``fs.cpSync`` but currently it's experimental so I'm using ``fs-extra``.
Can we have anyone who can test the cli on windows?

Comment (dai-shi) [reply contrib/create-waku/src/cli.ts]: Maybe @Sharlottes @hyf0

**Review by dai-shi — COMMENTED** [review]
Some more comments. I think we are getting close.

File: contrib/create-waku/src/cli.ts (hunk 26 above)
Comment (hyf0) [reply contrib/create-waku/src/cli.ts]: I'm back to mac world now. :-). Probably not able to be of help.

File: contrib/create-waku/src/cli.ts [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+  } catch (cancelled) {
+    if (cancelled instanceof Error) {
+      console.log(cancelled.message);
+    }
+    process.exit(1);
+  }
+
+  const { packageName, shouldOverwrite, chooseProject } = result;
+
+  const root = path.join(cwd, targetDir);
```
[the remark this answers is not in the document]
Comment (vasucp1207) [reply contrib/create-waku/src/cli.ts:101]: For local development, cwd here is for the ``create-waku`` dir, ``template`` folder is copied under this.

[the remark this answers is not in the document]
Comment (dai-shi) [reply contrib/create-waku/src/cli.ts:101]: We may want to try how it works with `npx`, but my assumption is cwd is where people run the command. So, it has to be something like this:
```js
const template = path.join(__dirname, "template");
const root = path.resolve(targetDir);
```

[the remark this answers is not in the document]
Comment (vasucp1207) [reply contrib/create-waku/src/cli.ts:101]: Yeah, I get the point of using the ``__dirname`` this is relative to the script location.

File: contrib/create-waku/src/cli.ts [hunk 28]
```diff
@@ -0,0 +1,191 @@
...
+      pnpm: "pnpm dev",
+      yarn: "yarn dev",
+      npm: "npm run dev",
+    },
+  };
+
+  console.log(`\nDone. Now run:\n`);
+  if (root !== cwd) {
+    console.log(`${bold(green(`cd ${path.relative(cwd, root)}`))}`);
+  }
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:183]: I don't think the condition becomes `false` in general:
```suggestion
  console.log(`${bold(green(`cd ${path.relative(cwd, root)}`))}`);
```

File: contrib/create-waku/src/cli.ts [hunk 29]
```diff
@@ -0,0 +1,191 @@
...
+      existingPackageJsonPath,
+      JSON.stringify({
+        ...existingPackageJson,
+        name: packageName ?? toValidPackageName(targetDir),
+        version: "0.0.0",
+      }, null, 2)
+    );
+  };
+
+  render(chooseProject as string);
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:158]: It's probably inferred correctly.
```suggestion
  render(chooseProject);
```

File: contrib/create-waku/src/generateTemplates.test.ts [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+    true,
+    console.log(bold(green(`Template generated successfully.`)))
+  );
+
+  // Clean up the generated file after the test
+  fse.removeSync(targetDir);
+}
+
+// Run the test
+testTemplateGeneration();
```
Comment (dai-shi) [inline contrib/create-waku/src/generateTemplates.test.ts:30]: I don't think this test makes much sense. Let's remove it for now.
(I think we should add tests with Vitest, to cover each function, but it can be a separate PR.)

**Review by dai-shi — CHANGES_REQUESTED** [review]
Please check CI errors too.

File: contrib/create-waku/src/generateTemplates.test.ts (hunk 30 above)
Comment (vasucp1207) [reply contrib/create-waku/src/generateTemplates.test.ts:30]: Yeah, I actually do not decide how to test this, so a separate PR in future is a good idea.

**Comment by vasucp1207** [comment]
> Please check CI errors too.

I also run prettier, don't know why it fails

**Comment by dai-shi** [comment]
Maybe we need to add the empty config line
https://github.com/dai-shi/waku/blob/33c41546aac3be3d706f22a055589ef82398b0f4/package.json#L93
in create-waku/package.json too.

File: contrib/create-waku/src/cli.ts [hunk 31]
```diff
@@ -0,0 +1,171 @@
...
+
+  const pkg = {
+    name: packageName ?? toValidPackageName(targetDir),
+    version: "0.0.0",
+  };
+
+  console.log("Setting up project...");
+  const templateRoot = path.join(__dirname, "../template");
+
+  const render = function render(templateName: string) {
```
Comment (dai-shi) [inline contrib/create-waku/src/cli.ts:117]: Can we avoid the function? I don't necessarily prefer a long function (`init` in this case), and we generally should split into smaller functions (better for testing), but this doesn't seem necessary and `render` sounds confusing (we just copy and patch json).

**Review by dai-shi — COMMENTED** [review]
Probably the last review comment.

File: contrib/create-waku/src/cli.ts (hunk 31 above)
Comment (vasucp1207) [reply contrib/create-waku/src/cli.ts]: Removed `render` as it is not necessary, splitting `init` in the same file or creating another file for the functions.

**Review by dai-shi — APPROVED** [review]
The code looks good to me.
I think we want to make the project monorepo at this point. I need to learn how to do it. Give me some time.

**Comment by dai-shi** [comment]
@vasucp1207 #122 is merged. Please follow the new structure.

**Comment by dai-shi** [comment]
Please check CI errors.

File: packages/create-waku/package.json [hunk 32]
```diff
@@ -1,10 +1,29 @@
 {
   "name": "create-waku",
   "version": "0.4.10",
-  "author": "Daishi Kato",
-  "repository": {
-    "type": "git",
-    "url": "https://github.com/dai-shi/waku.git"
+  "bin": "dist/cli.js",
```
Comment (dai-shi) [inline packages/create-waku/package.json:4]:
```suggestion
  "author": "Daishi Kato",
  "contributors": [
    "Vasu Singh"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/dai-shi/waku.git"
  },
  "bin": "./dist/cli.js",
```

File: packages/create-waku/package.json [hunk 33]
```diff
@@ -2,9 +2,36 @@
...
+    "prompts": "^2.4.2"
+  },
+  "devDependencies": {
+    "@types/fs-extra": "^11.0.1",
+    "@types/node": "^20.4.7",
+    "@types/prompts": "^2.4.4",
+    "esbuild": "0.18.17",
+    "typescript": "^5.1.6"
+  },
+  "prettier": {}
```
Comment (dai-shi) [inline packages/create-waku/package.json:36]: Maybe we don't need it because we have in the repo root?
```suggestion
  }
```

File: packages/create-waku/src/cli.ts [hunk 34]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  const { packageName, shouldOverwrite, chooseProject } = result;
+
+  const root = path.resolve(targetDir);
+
+  if (shouldOverwrite) {
+    fse.emptyDirSync(root);
+  } else if (!fs.existsSync(root)) {
+    fs.mkdirSync(root);
```
Comment (dai-shi) [inline packages/create-waku/src/cli.ts:107]: Curious if we should have the recursive option:
https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#fsmkdirsyncpath-options

File: packages/create-waku/package.json [hunk 35]
```diff
@@ -2,9 +2,36 @@
...
+  "bin": "./dist/cli.js",
+  "files": [
+    "src",
+    "dist",
+    "template"
+  ],
+  "type": "commonjs",
+  "scripts": {
+    "start": "node dist/cli.js",
+    "template": "cp -r ../../examples template/",
```
Comment (dai-shi) [inline packages/create-waku/package.json:21]: We should add `/packages/create-waku/template` in the .gitignore file.

**Review by dai-shi — CHANGES_REQUESTED** [review]
Please check CI errors again.

File: packages/create-waku/src/cli.ts (hunk 34 above)
Comment (vasucp1207) [reply packages/create-waku/src/cli.ts]: Yes, good for creating nested dir.

File: packages/create-waku/package.json [hunk 36]
```diff
@@ -2,9 +2,35 @@
...
     "url": "https://github.com/dai-shi/waku.git"
   },
-  "bin": "./cli.js"
+  "bin": "./dist/cli.js",
+  "files": [
+    "src",
+    "dist",
+    "template"
+  ],
+  "type": "commonjs",
```
Comment (dai-shi) [inline packages/create-waku/package.json:18]: Future TODO: make type=module and use swc instead of esbuild.

File: packages/create-waku/src/cli.ts [hunk 37]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env node
+
+// FIXME the following is temporarily and will remove it soon
+/* eslint-disable */
+
```
Comment (dai-shi) [inline packages/create-waku/src/cli.ts:5]: My intention is to remove this in this PR.

```suggestion
```

File: packages/create-waku/package.json (hunk 36 above)
Comment (vasucp1207) [reply packages/create-waku/package.json:18]: Sure, I will take look into this.

**Review by dai-shi — APPROVED** [review]
Let's merge it. Feel free to work on improvements later.

**Comment by dai-shi** [comment]
https://www.npmjs.com/package/create-waku/v/0.5.1 is published.
@vasucp1207 Can you please check if it works as expected?
